contracts/stellar-interchain-token-service/src/migrate.rs

```diff
@@ -70,68 +30,44 @@ pub mod legacy_storage {
 }
 
 impl CustomMigratableInterface for InterchainTokenService {
-    type MigrationData = (
-        BytesN<32>, /* new_wasm_hash */
-        Vec<(BytesN<32> /* token_id */, u64 /* epoch */)>,
-    );
+    type MigrationData = CustomMigrationData;
     type Error = ContractError;
 
     fn __migrate(env: &Env, migration_data: Self::MigrationData) -> Result<(), Self::Error> {
-        let (new_wasm_hash, token_ids_and_epochs) = migration_data;
-
-        for (token_id, epoch) in token_ids_and_epochs {
-            let token_id_config = storage::try_token_id_config(env, token_id.clone())
+        let CustomMigrationData {
+            new_token_manager_wasm_hash,
+            new_interchain_token_wasm_hash,
+            token_ids,
+            current_epoch,
+        } = migration_data;
+
+        for token_id in token_ids.into_iter() {
+            let TokenIdConfigValue {
+                token_address: interchain_token,
+                token_manager,
+                ..
+            } = storage::try_token_id_config(env, token_id.clone())
                 .ok_or(ContractError::InvalidTokenId)?;
 
-            Self::migrate_token_manager(env, token_id_config.token_manager, &new_wasm_hash)?;
-            Self::migrate_interchain_token(env, &token_id, &new_wasm_hash)?;
-            Self::migrate_flow_key(env, &token_id, epoch)?;
-        }
+            UpgradableClient::new(env, &token_manager).upgrade(&new_token_manager_wasm_hash);
+            UpgradableClient::new(env, &interchain_token).upgrade(&new_interchain_token_wasm_hash);
 
-        Ok(())
-    }
-}
-
-impl InterchainTokenService {
-    fn migrate_token_manager(
-        env: &Env,
-        token_manager_address: Address,
-        new_wasm_hash: &BytesN<32>,
-    ) -> Result<(), ContractError> {
-        UpgradableClient::new(&env, &token_manager_address).upgrade(new_wasm_hash);
-
-        storage::set_token_manager_wasm_hash(env, new_wasm_hash);
-
-        Ok(())
-    }
+            let flow_key = legacy_storage::FlowKey {
+                token_id: token_id.clone(),
+                epoch: current_epoch,
```

Comment:
compute the `current_epoch` via the helper instead of passing it as an arg. It's more robust this way.